awx_collection/plugins/lookup/tower_api.py

```diff
@@ -0,0 +1,168 @@
+# (c) 2020 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = """
+lookup: tower_api
+author: John Westcott IV (@john-westcott-iv)
+short_description: Search the API for objects
+requirements:
+  - None
+description:
+  - Returns GET requests from the Ansible Tower API. See
+    U(https://docs.ansible.com/ansible-tower/latest/html/towerapi/index.html) for API usage.
+extends_documentation_fragment:
+  - awx.awx.auth_plugin
+options:
+  _terms:
+    description:
+      - The endpoint to query, i.e. teams, users, tokens, job_templates, etc.
+    required: True
+  query_params:
+    description:
+      - The query parameters to search for in the form of key/value pairs.
+    type: dict
+    required: False
+    aliases: [query, data, filter, params]
+  expect_objects:
+    description:
+      - Error if the response does not contain either a detail view or a list view.
+    type: boolean
+    default: False
+    aliases: [expect_object]
+  expect_one:
+    description:
+      - Error if the response contains more than one object.
+    type: boolean
+    default: False
+  return_objects:
+    description:
+      - If a list view is returned, promote the list of results to the top-level of list returned.
+      - Allows using this lookup plugin to loop over objects without additional work.
+    type: boolean
+    default: True
+  return_all:
+    description:
+      - If the response is paginated, return all pages.
+    type: boolean
+    default: False
+  return_ids:
+    description:
+      - If response contains objects, promote the id key to the top-level entries in the list.
+      - Allows looking up a related object and passing it as a parameter to another module.
+    type: boolean
+    aliases: [return_id]
+    default: False
+  max_objects:
+    description:
+      - if C(return_all) is true, this is the maximum of number of objects to return from the list.
+    type: integer
+    default: 1000
+
+notes:
+  - If the query is not filtered properly this can cause a performance impact.
+  - In addition, the built in threshold is 10,000 items; if the query returns more an exception will be thrown.
+"""
+
+EXAMPLES = """
+- name: Load the UI settings
+  debug:
+    msg: "{{ query('awx.awx.tower_api', 'settings/ui') }}"
+
+- name: Lookup any users who are admins
+  debug:
+    msg: "{{ query('awx.awx.tower_api', 'users', query_params={ 'is_superuser': true }) }}"
```

Comment:
I ran through these examples with your branch, and I would like to add more elaboration.

```yaml
    - name: Load the UI settings  # plugin returns a list of 1 but lookup returns first element
      debug:
        msg: "{{ lookup('awx.awx.tower_api', 'settings/ui') }}"

    - name: Lookup any users who are admins  # use query to return the full list
      debug:
        msg: "{{ query('awx.awx.tower_api', 'users', query_params={ 'is_superuser': true }) | map(attribute='username') }}"

    # use lookup to return first item
    # the filter may return more than 1 item, it is users responsibility to assure this item is what they want
    # this can be used to pass related items to modules in the collection
    - name: Lookup a users ID
      debug:
        msg: "{{ lookup('awx.awx.tower_api', 'users', query_params={ 'username': 'admin' }, return_ids=True) }}"

    - name: debug all organizations in a loop  # use query to return a list
      debug:
        msg: "Organization description={{ item['description'] }} id={{ item['id'] }}"
      loop: "{{ query('awx.awx.tower_api', 'organizations') }}"
      loop_control:
        label: "{{ item['name'] }}"
```

This is probably going to be our primary blurb of documentation that users make use of, so we should make it count.

Reply:
Yes, I would like to expand this section to include examples of all of the parameters. In addition, I'd like to review the tests to make sure we have corresponding tests for every option.

Reply:
it might seem trite, but I also think it would be nice to have a line continuation example, for instance:

```yaml
  - name: Debug using the rrule plugin
    debug:
      msg: >
        {{ query(
          'awx.awx.tower_schedule_rrule',
          'minute',
          start_date='2020-4-16 03:45:07',
          end_on='2'
        ) }}
```

That's closer to how I would like to see a playbook actually written.